### ISSUE TITLE
When events or aggregates are decoded, nested maps have all keys converted to atoms

### ISSUE THREAD

**Issue by Fire-Dragon-DoL** [commanded/commanded#301]
The issue is due to the following line: https://github.com/commanded/commanded/blob/v0.19.1/lib/commanded/serialization/json_serializer.ex#L24

While is not directly a bug, it's a great vector of DoS, since it will generate atoms for any data inside an event.
Data inside events it's usually user-provided and while some might fit a specific schema, other data might be more dynamic (custom variables of some kind).

In my specific case, I have a field called `custom` that is a map of strings to strings, where the key is a column header from a csv file and the value is specific to a row in the csv (there is one instance of aggregate per row in the csv). I can't know upfront all the columns of the CSV file and any column is allowed (a minimum pre-defined set is required).

Currently, if an event has a nested map, that map will always be decoded to atom keys and the behavior can't be circumvented without changing Commanded.

I suggest that instead, all keys will be decoded to strings, than all the top level keys will be converted to atoms and finally convert everything to a struct. In this way, nested maps will be keys and if the developer needs them as atom, it can use the `JsonDecoder` protocol.

**Comment by Fire-Dragon-DoL** [comment]
I can provide a PR between Monday and Wednesday

**Comment by slashdotdash** [comment]
This is covered in issue #299 to allow options to be passed to the configured serializer (such as the default JSON serializer).

**Comment by Fire-Dragon-DoL** [comment]
Hey @slashdotdash I looked carefully at #299 before writing this issue and I don't think it addresses that _entirely_.
If `maps: :strings` is passed, the `Struct` when decoded will have no keys filled and the user is forced to do the transformation entirely.

My suggestion was to use `maps: :strings` but add a mechanism that still transforms the default map keys to atoms, just not recursively (it shouldn't break compatibility either in this way).

Then #299 is welcome obviously, but at that point it might be easier to have just 2 configurable functions (or a module implementing some callbacks) so that the flexibility is maximum.
If only the options to Jason are allowed, the behavior I'm describing still cannot be implemented.

Let me know of your preferences, I'm still open to create such PR (covering #299 too)

Cheers and thanks for following through the issue!

**Comment by Fire-Dragon-DoL** [comment]
@slashdotdash I figured that the `serialized` option of the `commanded` application already addresses it. I can even use it to configure the options for Jason.
I basically copied the existing implementation with the changes needed to avoid parsing json into atoms. Unfortunately some code relied on that (bad) so I had to go in and change all the places which relied on parsing directly to atoms the JSON.

Maybe pointing this behavior in the README or the Guides might be important, since effectively decoding from JSON to atom keys is dangerous
